### PR TITLE
Add Raft Autopilot Configuration Resource

### PR DESCRIPTION
**Pull request by mbillow**
Adding a resource for configuring Vault's Raft Autopilot.

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #0000

Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-vault/blob/master/CHANGELOG.md):

```release-note
- `resource/vault_raft_autopilot`: Configure Vault's Raft Autopilot settings
```

Output from acceptance testing:

```
$ TESTARGS="--run RaftAutopilotConfig" make testacc
...
=== RUN   TestAccRaftAutopilotConfig_basic
--- PASS: TestAccRaftAutopilotConfig_basic (52.89s)
PASS
ok      github.com/hashicorp/terraform-provider-vault/vault     54.953s

```

### CODE REVIEW

File: vault/resource_raft_autopilot.go [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+	"github.com/hashicorp/vault/api"
+)
+
+var autopilotPath = "sys/storage/raft/autopilot/configuration"
+var autopilotDefaults = map[string]interface{}{
+	"cleanup_dead_servers":               false,
+	"dead_server_last_contact_threshold": "24h0m0s",
+	"last_contact_threshold":             "10s",
+	"max_trailing_logs":                  1000,
+	"min_quorum":                         0,
```
Comment (vinay-gopalan) [inline vault/resource_raft_autopilot.go:17]: Could we change this to match the default value to 3 as [listed here](https://www.vaultproject.io/api-docs/system/storage/raftautopilot#min_quorum)?

File: vault/resource_raft_autopilot.go [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+	"server_stabilization_time":          "10s",
+}
+
+func raftAutopilotConfigResource() *schema.Resource {
+	fields := map[string]*schema.Schema{
+		"cleanup_dead_servers": {
+			Type:        schema.TypeBool,
+			Description: "Specifies automatic removal of dead server nodes periodically.",
```
Comment (vinay-gopalan) [inline vault/resource_raft_autopilot.go:25]:
```suggestion
			Description: "Specifies whether to remove dead server nodes periodically or when a new server joins. This requires that min-quorum is also set."
```

File: vault/resource_raft_autopilot.go [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+			Default:     autopilotDefaults["cleanup_dead_servers"],
+			Optional:    true,
+		},
+		"dead_server_last_contact_threshold": {
+			Type:        schema.TypeString,
+			Description: "Limit the amount of time a server can go without leader contact before being considered failed.",
```
Comment (vinay-gopalan) [inline vault/resource_raft_autopilot.go:31]:
```suggestion
			Description: "Limit the amount of time a server can go without leader contact before being considered failed. This takes effect only when cleanup_dead_servers is set."
```

File: vault/resource_raft_autopilot.go [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+		},
+		"max_trailing_logs": {
+			Type:        schema.TypeInt,
+			Description: "Maximum number of log entries in the Raft log that a server can be behind its leader before being considered unhealthy.",
+			Default:     autopilotDefaults["max_trailing_logs"],
+			Optional:    true,
+		},
+		"min_quorum": {
+			Type:        schema.TypeInt,
+			Description: "Minimum number of servers allowed in a cluster before autopilot can prune dead servers.",
```
Comment (vinay-gopalan) [inline vault/resource_raft_autopilot.go:49]:
```suggestion
			Description: "Minimum number of servers allowed in a cluster before autopilot can prune dead servers. This should at least be 3. Applicable only for voting nodes.",
```

File: website/docs/r/raft_autopilot.html.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+  min_quorum = 3
+  server_stabilization_time = "10s"
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+- `cleanup_dead_servers` – (Optional) Specifies automatic removal of dead server
```
Comment (vinay-gopalan) [inline website/docs/r/raft_autopilot.html.md:33]:
```suggestion
- `cleanup_dead_servers` – (Optional) Specifies whether to remove dead server nodes periodically or when a new server joins. This requires that `min-quorum` is also set.
```

File: website/docs/r/raft_autopilot.html.md [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+## Argument Reference
+
+The following arguments are supported:
+
+- `cleanup_dead_servers` – (Optional) Specifies automatic removal of dead server 
+nodes periodically.
+
+- `dead_server_last_contact_threshold` - (Optional) Limit the amount of time a 
+server can go without leader contact before being considered failed.
```
Comment (vinay-gopalan) [inline website/docs/r/raft_autopilot.html.md:37]:
```suggestion
server can go without leader contact before being considered failed. This takes effect only when `cleanup_dead_servers` is set.
```

File: website/docs/r/raft_autopilot.html.md [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+- `last_contact_threshold` - (Optional) Limit the amount of time a server can go 
+without leader contact before being considered unhealthy.
+
+- `max_trailing_logs` - (Optional) Maximum number of log entries in the Raft log 
+that a server can be behind its leader before being considered unhealthy.
+
+- `min_quorum` - (Optional) Minimum number of servers allowed in a cluster before 
+autopilot can prune dead servers.
```
Comment (vinay-gopalan) [inline website/docs/r/raft_autopilot.html.md:46]:
```suggestion
autopilot can prune dead servers. This should at least be 3. Applicable only for voting nodes.
```

**Review by vinay-gopalan — CHANGES_REQUESTED** [review]
Thanks for the PR! This looks good, just a few comments on some doc string updates and default value changes :)

**Comment by mbillow** [comment]
@vinay-gopalan I’ll update the default and push when I get back to my desk this afternoon. The documentation updates look good as well. I did just copy and paste them from the existing Vault docs though, so it might be good to also make these same suggestions there as well.

File: vault/resource_raft_autopilot.go (hunk 1 above)
Comment (mbillow) [reply vault/resource_raft_autopilot.go]: Updated ✅

**Comment by mbillow** [comment]
@vinay-gopalan Requested changes have been made. Latest acceptance test results:
```
=== RUN   TestAccRaftAutopilotConfig_basic
--- PASS: TestAccRaftAutopilotConfig_basic (49.49s)
PASS
ok      github.com/hashicorp/terraform-provider-vault/vault     50.857s
```

**Review by vinay-gopalan — APPROVED** [review]

**Pull request merged**